bob-backend/src/pearl/group.rs

```diff
@@ -333,11 +344,14 @@ impl Group {
     }
 
     pub async fn attach(&self, start_timestamp: u64) -> BackendResult<()> {
-        let holders = self.holders.read().await;
-        if holders
-            .iter()
-            .map(|x| x.start_timestamp())
-            .any(|timestamp| timestamp == start_timestamp)
+        if {
+            let holders = self.holders.read().await;
```

Comment:
`attach` is a rare operation, so I think it is better to acquire `write` lock at the beginning of the function and held it until the end. `save_pearl` can be manually inlined here to avoid deadlock